src/crimson/osd/pg.h

```diff
@@ -822,13 +827,18 @@ class PG : public boost::intrusive_ref_counter<
     return can_discard_replica_op(m, m.get_map_epoch());
   }
 
+  void set_pglog_based_recovery_op(PglogBasedRecovery *op) final;
+  void reset_pglog_based_recovery_op() final;
+  void cancel_pglog_based_recovery_op();
+
 private:
   // instead of seastar::gate, we use a boolean flag to indicate
   // whether the system is shutting down, as we don't need to track
   // continuations here.
   bool stopping = false;
 
   PGActivationBlocker wait_for_active_blocker;
+  PglogBasedRecovery* pglog_based_recovery_op = nullptr;
```

Comment:
Can we switch to std::optional?

Reply:
Um, `pglog_based_recovery_op` is not owned by `PG` but by `PerShardState::registry`, so it seems it can't be std::optional. Am I right?